File: 14 Longest Common Prefix/14 Longest Common Prefix.py
"""
Write a function to find the longest common prefix string amongst an array of strings.
If there is no common prefix, return an empty string "".

Example 1:

Input: strs = ["flower","flow","flight"]
Output: "fl"

Example 2:

Input: strs = ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.
"""
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        length=201
        for i in range(len(strs)):
            if strs[i]=="":
                return ""
            if len(strs[i])<length:
                minindex=i
                length=len(strs[i])
        candidate=strs.pop(minindex)
        for i in range(len(candidate),0,-1):
            found=True
            candidate=candidate[:i]
            for j in range(0,len(strs)):
                if strs[j][:i]!=candidate:
                    found=False
                    break
            if found:
                return candidate
        if not found:
            return ""

File: 14 Longest Common Prefix/14 Longest Common Prefix.go
func longestCommonPrefix(strs []string) string {
    prefix := ""
	j := 0
	if len(strs) == 0 {
		return ""
	}
	if len(strs) == 1 {
		return strs[0]
	}
	for {
		tmp := make(map[string]int)
		for _, element := range strs {
			word := element
			if j > len(word) || len(word) == 0 {
				return prefix
			} else {
				if j == len(word) {
					tmp[word[j:]] +=1
				} else{
					tmp[word[j:j+1]] += 1
				}

				if len(tmp) > 1 {
					return prefix
				}
			}
		}
		if len(tmp) != 1 {
			return prefix
		} else {
			for k := range tmp {
				prefix = prefix + k
			}
		}
		j += 1
	}

}
